pluma/io/pupil.py

```diff
@@ -0,0 +1,48 @@
+import struct
+import warnings
+import numpy as np
+
+from typing import Union
+
+import pandas as pd
+import os
+from pluma.io.path_helper import ComplexPath, ensure_complexpath
+
+
+def load_pupil(filename: str,
+               dtypes: list[list[tuple[str, type]]],
```

Comment:
I was wondering why this had to be a list of lists of tuples, and I think I now understand given the documentation of `np.dtype`, which unfortunately does not accept directly a dictionary as a "list of tuples".

Another option for perhaps making the function spec more readable would be to pass the lists of tuples as either `**kwargs` or named arguments, so we could write something like the following:

```python
    streams.PupilLabs.Data.Gaze =                 PupilStream(
        'PupilLabs/Gaze',
        frame0=[('SensorId', np.string_, 36)],
        frame1=[('Timestamp', np.uint64)],
        frame2=[('GazeX', np.single), ('GazeY', np.single)],
        device='PupilLabs',
        streamlabel='PupilLabs.Gaze.Data',
        root=root, autoload=autoload,
        parent_dataset=parent_dataset)
```

There should be no functional difference hopefully, but it would act as a reminder of the internal data format of these streams, and of which bits go into decoding which frames.

Reply:
I agree that the list[list[tuple]] makes things a little less readable than named arguments. The thought process behind this is:

- Eventually, we can probably make a generic NetMQ reader that works for both the Glia and pupil streams. These two streams have different message protocols (Pupil always 3 frames, Glia has 2 frames) but by using the data type list we can arbitrarily decode messages with varying numbers of frames per sample/message. Using optional arguments for each frames means that each stream with a different NetMQ message protocol needs to be defined separately.
- **kwargs is an alternative, but personally I like the certainty that the length of the outer list of dtype should reflect the number of frames per message. If other parameters get added to **kwargs later that are not related to the data types of each frame then the number of arguments doesn't reflect the frames per message.
- The length of the outer list is also used at the moment to decide if any of the frame bin files contain jagged / variable length data (e.g. raw h264 encoded camera frames). Admittedly this is quite specific to the sensor at the moment, but in future could be more generalised with some flag in the list.

Reply:
This makes sense, thanks for the clarification. I still think there are two possible levels here: 1. the schema specification level; and 2. the reader implementation level.

So I guess the way I probably would have done it is to have a generic abstract class for NetMQ messages that follows your generic spec, and then have two derived implementations for Pupil and Glia where the specs for them are fixed in the class itself.

That way we could reduce duplication in both code flow and in the schema.